core/src/main/java/org/apache/iceberg/SnapshotParser.java

```diff
@@ -153,35 +139,22 @@ static Snapshot fromJson(FileIO io, JsonNode node) {
 
     Integer schemaId = JsonUtil.getIntOrNull(SCHEMA_ID, node);
 
-    if (node.has(MANIFEST_LIST)) {
```

Comment:
This is still needed. If Iceberg is reading an old table that embeds the manifest list locations, then there's no reason why it should fail. That would be a backward-incompatible change. As long as we don't write out a form with the manifests, we should be fine.

Reply:
Okay, I took a deeper look into our options here. I don't think that we should break tables that have embedded list of manifests. I'm okay with no longer writing the embedded list, but I don't want to break compatibility with older tables.

That means we need this check.

Reading should work fine by lazily converting to `GenericManifestFile` when a `FileIO` is provided through the `Snapshot` API. However, making changes to the table metadata would break when writing out old snapshots. I was thinking earlier that we could rely on `SnapshotProducer` to rewrite, but if the update is a metadata change then we don't get a new snapshot that uses a list. We could have an operation that rewrites the snapshots to use lists, but that seems like a lot of work.

Instead, I think that the cleanest option is to pass `FileIO` to `allManifests` for the write path where `v1ManifestLocations` is being called. We can either do that by passing in a real `FileIO` or by creating a fake `FileIO`.

I'd use a fake `FileIO` because we know that it is only used to create `InputFile` instances for the list of manifests and then only `path` is called on those input files to return the location. I think that's a clean way to remove `FileIO` without adding a new method to the `Snapshot` interface. It's also pretty small:

```java
  private static class DummyFileIO implements FileIO {
    @Override
    public InputFile newInputFile(String path) {
      return new InputFile() {
        @Override
        public long getLength() {
          throw new UnsupportedOperationException();
        }

        @Override
        public SeekableInputStream newStream() {
          throw new UnsupportedOperationException();
        }

        @Override
        public String location() {
          return path;
        }

        @Override
        public boolean exists() {
          return true;
        }
      };
    }

    @Override
    public OutputFile newOutputFile(String path) {
      throw new UnsupportedOperationException();
    }

    @Override
    public void deleteFile(String path) {
      throw new UnsupportedOperationException();
    }
  }
```

I think I prefer our options in this order:
1. Use a fake FileIO to get the files
2. Keep passing a real FileIO in the write path only
3. Add `v1ManifestLocations` to the `Snapshot` interface

@nastra, what do you think?

Reply:
Using a fake FileIO in `for (ManifestFile file : snapshot.allManifests(DUMMY_FILE_IO))` sounds great and we don't need expose the `v1ManifestLocations` on the `Snapshot` API itself. I went ahead and implemented that and pushed that to https://github.com/apache/iceberg/pull/5734.

I would suggest that we first finish up https://github.com/apache/iceberg/pull/5734 and then come back to this PR to deal with the `write.manifest-lists.enabled` flag (aka deprecating it and removing it from the write path, but still keeping it in the read path in `SnapshotParser` as you indicated above). I will update this PR here accordingly once https://github.com/apache/iceberg/pull/5734 is in